text/0000-cargo-test-coverage.md

```diff
@@ -0,0 +1,343 @@
+flag to all crates within the dependency graph, including the `regex` crate and any transitive dependencies.
+
+This should not break any existing workflows and is strictly an opt-in feature.
+
+To use this new feature do the following:
+
+`cargo test --coverage`
+
+This flag would also be responsible for setting the `LLVM_PROFILE_FILE` environment variable which is read by LLVM
+to generate a `.profraw` file for each test executable that is run. Once again if the environment variable is already set,
+then Cargo would not make any changes and would leave the value as is to use the user-defined file name. If the environment
+variable is not set, Cargo would set it to ensure a unique naming scheme is used for each `.profraw` file that would be
+generated.
+
+Once the tests have finished running, Cargo would leverage LLVM tooling to analyze profraw files and generate
+coverage reports for a Rust developer. These tools would be `llvm-profdata` and `llvm-cov` and can be used by
+adding the `llvm-tools-preview` component to the current toolchain, `rustup component add llvm-tools-preview`.
+This would be a requirement to use this new feature since these are the tools necessary to analyze coverage results
+and generate a code coverage report. If the component is not installed for the current toolchain, an error will occur
+with a message stating the the `llvm-tools-preview` component is required to generate a code coverage report.
+
+For example, a Rust developer can invoke the following Cargo command to generate an HTML coverage report
+for the current crate or `default-members` in a workspace:
+
+```
+cargo test --coverage
+```
+
+This will generate a code coverage report in HTML format in the `target/coverage/` directory. If a workspace
+does not have `default-members`, then all members would be instrumented.
+
+Run the following cargo CLI to choose a specific set of crates to instrument and override the defaults:
+
+```
+cargo test --coverage:crate1,crate3,foo
+```
+
+To override the default output format and directory, the `--coverage-format` and `--coverage-output-directory`
+can be passed to the cargo CLI:
+
+```
+cargo test --coverage --coverage-format=lcov --coverage-output-directory=src/coverage
+```
+
+This will generate a code coverage report in lcov format in the `src/coverage/` directory.
+
+The supported options for the `--coverage-format` option are:
+
+1. `html`
+2. `lcov`
+3. `json`
+
+These updates to Cargo would be sufficient enough to ensure that a Rust developer would have control over what crates are
+instrumented, the format in which code coverage results are generated and where they are stored. This would also allow
+the Rust developer to no longer have to set environment variables manually to ensure crates are instrumented
+for gathering coverage data and can generate code coverage results with a single command.
+
+# Drawbacks
+[drawbacks]: #drawbacks
```

Comment:
_Disclaimer all my knowledge comes from spending an unhealthy amount of time looking at LLVM coverage instrumentation code, some inaccuracies may exist when talking about other backends._ So cranelift currently has no coverage instrumentation https://github.com/bytecodealliance/wasmtime/issues/1151 so support for that can't really be done. For libgccjit used by rustc_codegen_gcc I'm not too eu fait but there's https://github.com/rust-lang/rustc_codegen_gcc/blob/master/src/coverageinfo.rs which looks to work the same as LLVM with inserting counters at source locations. So the part of the RFC regarding filtering which crates counters are added to should be directly applicable. Depending where in rustc it's called it may be able to reuse the code as well.

LLVM_PROFILE_FILE isn't applicable but tbh rustc could set a "sensible default value" for any variables like this that avoid profile naming clashes and mean the user doesn't have to touch it for the common usecase `cargo test --coverage`

Reply:
That's a fair point. Currently this RFC notes that the proposed feature specifically targets the LLVM codegen backend. I can add this as a note in the drawbacks section and also add a note of supporting other backends in the `Future Possibilities` section.